lib/src/Uncertainty/Distribution/Multinomial.cxx

```diff
@@ -136,49 +136,86 @@ Point Multinomial::getRealization() const
+     * the i-th component is generated using a Binomial distribution */
+    n = n_;
+    sum = 1.0;
+    for (UnsignedInteger i = 0; i < dimension; ++i)
+    {
+      /* The current component follow a binomial distribution with parameters p_[i] and N */
+      xI = DistFunc::rBinomial(n, p_[i] / sum);
+      sample(k, i) = xI;
+      n -= xI;
+      if (n == 0) break;
+      sum -= p_[i];
+    }
+  }
+  return sample;
+}
+
 /* Get the PDF of the distribution */
-Scalar Multinomial::computePDF(const Point & point) const
+Scalar Multinomial::computePDF(const Indices & point) const
 {
   const UnsignedInteger dimension = getDimension();
-  if (point.getDimension() != dimension) throw InvalidArgumentException(HERE) << "Error: the given point must have dimension=" << dimension << ", here dimension=" << point.getDimension();
+  if (point.getSize() != dimension) throw InvalidArgumentException(HERE) << "Error: the given point must have dimension=" << dimension << ", here dimension=" << point.getSize();
 
   // First, check the validity of the input
-  Scalar sumX = 0.0;
+  UnsignedInteger sumX = 0.0;
   for (UnsignedInteger i = 0; i < dimension; ++i)
   {
-    const Scalar k = point[i];
+    const UnsignedInteger k = point[i];
     // Early exit if the given point is not in the support of the distribution
-    if ((std::abs(k - round(k)) > supportEpsilon_) || (k < -supportEpsilon_) || (k > n_ + supportEpsilon_)) return 0.0;
+    // Hear we cheat a little bit: if x_i is negative, the conversion to an
+    // unsigned integer leads to a value well above the value of n_, allowing
+    // to test for x_i < 0 and x_i > n in one test
+    if (k > n_) return 0.0;
     sumX += k;
   }
-  if (sumX > n_ + supportEpsilon_) return 0.0;
-  if ((sumP_ >= 1.0) && (sumX < n_ - supportEpsilon_)) return 0.0;
-  Scalar logPDF = -1.0;
-  if (sumP_ < 1.0) logPDF = lgamma(n_ + 1.0) - lgamma(n_ - sumX + 1.0) + (n_ - sumX) * log1p(-sumP_);
-  // In the case sumP_ >= 1.0, the PDF is positive only if sumX == n_
-  else logPDF = lgamma(n_ + 1.0);
+  if (sumX > n_) return 0.0;
+  if ((sumP_ >= 1.0) && (sumX < n_)) return 0.0;
+  Scalar sumP = sumP_;
+  UnsignedInteger sumK = 0;
+  UnsignedInteger k = n_ - sumX;
+  Scalar pdf = 0.0;
+  // If the multinomial distribution has been defined as X_1+...+X_d<=N, add a X_0 with value N-(X_1+...+X_d) and probability 1-sumP
+  if (sumP < 1.0)
+  {
+    pdf = DistFunc::dBinomial(n_, 1.0 - sumP, k);
+    sumK += k;
+  }
   for (UnsignedInteger i = 0; i < dimension; ++i)
   {
-    const Scalar k = point[i];
-    // For p_[i] > 0, it is possible to obtain any value of k between 0 and n
-    if (p_[i] > 0.0) logPDF += k * std::log(p_[i]) - lgamma(k + 1.0);
-    // Else only k == 0 is allowed, with a zero contribution to the log PDF
-    else if (k > 0) return 0.0;
+    // Here we know that round(point[i]) >= 0
+    const UnsignedInteger k = static_cast<UnsignedInteger>(round(point[i]));
```

Comment:
k => warning: declaration shadows a local variable [-Wshadow]

Reply:
indeed k declared first as Scalar then UI...

Reply:
k is declared as a const UnsignedInteger in a limited scope (line 182) then as an UnsignedInteger (line 194) then as a const UnsignedInteger in a limited scope (line 205), where the shadowing occurs. It was a Scalar in the previous version of the code.
I fixed it by declaring k as an UnsignedInteger once and for all before the first loop.